### PR TITLE
fix: Stale series with ruler evaluation delay

### PR DESCRIPTION
**Pull request by simonswine**
**What this PR does**:

When ruler evaluation delay is enabled, the upstream rule evaluation logic adds staleness markers for missing series at the actual timestamp when the evaluation is run rather than the timestamp when the series was missing: (see:
https://github.com/prometheus/prometheus/blob/6c56a1faaaad07317ff585bda75b99bdba0517ad/rules/manager.go#L647-L660).

This can cause larger gaps than there actually are, as once the data becomes available again it can't be written as it would be out-of-order.

This fix shifts those staleness marker back in time by the evaluation delay set for the individual tenant.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by gotjosh** [comment]
cc: @gouthamve

File: integration/ruler_test.go
Comment (jdbaldry) [inline integration/ruler_test.go:191]: I'm not sure I understand the comment here

**Review by jdbaldry — COMMENTED** [review]
One typo

File: integration/ruler_test.go
Comment (jdbaldry) [reply integration/ruler_test.go:191]: Oh I realize you didn't add the comment, sorry.

Comment (simonswine) [reply integration/ruler_test.go:191]: I don't understand it, so it didn't touch it

Comment (simonswine) [reply integration/ruler_test.go]: I have removed it now

File: integration/ruler_test.go
Comment (simonswine) [reply integration/ruler_test.go]: Thanks  :clap:

File: CHANGELOG.md
Comment (jtlisi) [inline CHANGELOG.md:48]: **suggestion:** I think it would be good to make it clear what was fixed in addition to the symptom is solves.
```suggestion
* [BUGFIX] Ruler: Ensure the stale markers generated for evaluated rules respect the set evaluation delay. This will avoid issues with samples with NaN be persisted with timestamps set ahead of the next rule evaluation. #3687
```

**Review by jtlisi — APPROVED** [review]
LGTM, modulo on minor changelog nit

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:48]:
```suggestion
* [BUGFIX] Ruler: Ensure the stale markers generated for evaluated rules respect the configured `-ruler.evaluation-delay-duration`. This will avoid issues with samples with NaN be persisted with timestamps set ahead of the next rule evaluation. #3687
```

File: integration/ruler_test.go
Comment (pracucci) [inline integration/ruler_test.go:236]: Since it's greater than 1 it will wait for 2 evaluations. Is that expected or you wanted `e2e.Greater(0)`?

File: integration/ruler_test.go
Comment (pracucci) [inline integration/ruler_test.go:215]: `posStale := 2` may be clearer.

File: integration/ruler_test.go
Comment (pracucci) [inline integration/ruler_test.go:221]: Is this the same but more clear, or am I missing something?

```suggestion
		series.Samples[pos].Timestamp = e2e.TimeToMilliseconds(now.Add(-evaluationDelay).Add(pos * time.Second))
```

File: integration/ruler_test.go
Comment (pracucci) [inline integration/ruler_test.go:251]: I'm not sure I do understand this. Why you need to wait `samplesToSend` evaluations?

File: integration/ruler_test.go
Comment (pracucci) [reply integration/ruler_test.go:236]: Also, I believe this test can be flaky because of this check. What you want is to wait 1 more evaluation after the `Push()`, so ideally you should get the the current `cortex_prometheus_rule_evaluations_total` and `WaitSumMetric()` by that value + 1.

**Review by pracucci — COMMENTED** [review]
Thanks for fixing this annoying issue! I took a look if there's a cleaner way to implement the ruler evaluation delay at all (to avoid the "hack/fix" done in the appender in this PR) but I haven't found any, so LGTM.

Before merging, I would like you to take a look at the comment I've left. Thanks!

File: integration/ruler_test.go
Comment (simonswine) [reply integration/ruler_test.go]: Good point, I have implemented as suggest, `Push -> SumMetric -> WaitSumMetric(Greater(result))`

File: integration/ruler_test.go
Comment (simonswine) [reply integration/ruler_test.go]: That's a remnant from a time when I had two staleness periods :+1:

File: integration/ruler_test.go
Comment (simonswine) [reply integration/ruler_test.go]: I have added a comment to clarify:
```
	// Wait until all the pushed samples have been evaluated by the rule. This
	// ensures that rule results are successfully written even after a
	// staleness period.```

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
